frame/collective/src/lib.rs

```diff
@@ -355,30 +656,55 @@ impl<T: Trait<I>, I: Instance> Module<T, I> {
 		approved: bool,
 		seats: MemberCount,
 		voting: Votes<T::AccountId, T::BlockNumber>,
-		proposal: T::Hash,
-	) {
+		proposal_hash: T::Hash,
+		proposal: Option<<T as Trait<I>>::Proposal>,
+	) -> Weight {
+		let db = T::DbWeight::get();
+		let mut weight: Weight = 0;
 		if approved {
-			Self::deposit_event(RawEvent::Approved(proposal));
+			Self::deposit_event(RawEvent::Approved(proposal_hash));
 
 			// execute motion, assuming it exists.
-			if let Some(p) = ProposalOf::<T, I>::take(&proposal) {
+			if let Some(p) = proposal {
 				let origin = RawOrigin::Members(voting.threshold, seats).into();
-				let ok = p.dispatch(origin).is_ok();
-				Self::deposit_event(RawEvent::Executed(proposal, ok));
+				let result = p.dispatch(origin);
+				Self::deposit_event(RawEvent::Executed(proposal_hash, result.is_ok()));
+				// TODO: is zero a decent default here?
+				weight = weight.saturating_add(get_result_weight(result).unwrap_or(0)); // P1
 			}
 		} else {
 			// disapproved
-			ProposalOf::<T, I>::remove(&proposal);
-			Self::deposit_event(RawEvent::Disapproved(proposal));
+			Self::deposit_event(RawEvent::Disapproved(proposal_hash));
 		}
 
 		// remove vote
-		Voting::<T, I>::remove(&proposal);
-		Proposals::<T, I>::mutate(|proposals| proposals.retain(|h| h != &proposal));
+		ProposalOf::<T, I>::remove(&proposal_hash);
+		Voting::<T, I>::remove(&proposal_hash);
+		let mut proposal_len = T::MaxProposals::get() as Weight;
+		Proposals::<T, I>::mutate(|proposals| {
+			proposal_len = proposals.len() as Weight;
+			proposals.retain(|h| h != &proposal_hash)
+		});
+		weight.saturating_add(db.reads_writes(1, 3)) // `Voting`, `Proposals`, `ProposalOf`
+			.saturating_add(490_000 * proposal_len) // P2
 	}
 }
 
 impl<T: Trait<I>, I: Instance> ChangeMembers<T::AccountId> for Module<T, I> {
+	/// Update the members of the collective. Votes are updated and the prime is reset.
+	///
+	/// # <weight>
+	/// ## Weight
+	/// - `O(MP + N)`
+	///   - where `M` old-members-count (governance-bounded)
+	///   - where `N` new-members-count (governance-bounded)
+	///   - where `P` proposals-count
+	/// - DB:
+	///   - 1 storage read (codec `O(P)`) for reading the proposals
+	///   - `P` storage mutations for updating the votes (codec `O(M)`)
+	///   - 1 storage write (codec `O(N)`) for storing the new members
+	///   - 1 storage write (codec `O(1)`) for deleting the old prime
+	/// # </weight>
 	fn change_members_sorted(
```

Comment:
This doesn't check that new members doesn't exceed MaxMembers.

The doc should say that it must not exceed MaxMembers. and implementation should print error in case it happens IMO. or maybe exceeded members should just not be inserted.

Reply:
So you mean that we should adjust the `ChangeMembers` trait to allow for errors in the methods?

Reply:
oh indeed we can also do that not sure what is better:
* state in the doc that it shouldn't exceed MaxMembers, and ditch exceeders or just insert all of them and do `support::print("Error....")`.
* change the signature of change_members_sorted. But I'm not how user of this trait will be able to handle error.

Reply:
Yeah, not sure what the expectations for interacting pallets are here.